### ISSUE TITLE
Ability to add dependency using a Provider type

### ISSUE THREAD

**Issue by mkobit** [gradle/gradle#3251]
As a plugin developer, I would like to provide lazy calculated values as dependencies. I could also use this as a build script author.

### Current Behavior

Exception thrown by configuration if a `Provider` type is passed in.

### Context

Mapping extension values or any other values to dependencies has to be done in an `afterEvaluate` type step right now.

### Steps to Reproduce (for bugs)

```groovy
tasks.wrapper.gradleVersion = '4.3-rc-2'

final myProp = objects.property(String)
myProp.set('com.google.guava:guava:23.0')

repositories {
  jcenter()
}

configurations {
  myConfig
}

dependencies {
  myConfig(myProp)
}
```

Fails with

```
> Cannot convert the provided notation to an object of type Dependency: value: com.google.guava:guava:23.0.                    
  The following types/formats are supported:                   
    - Instances of Dependency. 
    - String or CharSequence values, for example 'org.gradle:gradle-core:1.0'.                                                 
    - Maps, for example [group: 'org.gradle', name: 'gradle-core', version: '1.0'].                                            
    - FileCollections, for example files('some.jar', 'someOther.jar').                                                         
    - Projects, for example project(':some:project:path').     
    - ClassPathNotation, for example gradleApi().  
```

### Your Environment

* version 4.3-rc-2

**Comment by bmuschko** [comment]
The error message you are seeing is expected as we don't support assigning `Provider` types as dependencies. Can you give us a concrete example when you want to defer the evaluation of a dependency version? Is the use case that you want to be able to assign a dependency value through a property exposed by an extension?

**Comment by mkobit** [comment]
The error message was expected by me as `Provider` is still really new, and I see you guys doing a lot of recent work on it . I should remove the _"Steps to Reproduce (for bugs)"_ because this is definitely not a bug.

You have the most concrete case, like you said, is to map an extension property to a later calculated dependency

```groovy
myExtension {
  packageVersion = '1.2'
}
```

Right now, I believe I have to map this in an `afterEvaluate` ([here](https://github.com/mkobit/jenkins-pipeline-shared-libraries-gradle-plugin/blob/5acecf593c661a5c1daaa0dc0b0fcd9572705984/src/main/kotlin/com/mkobit/jenkins/pipelines/SharedLibraryPlugin.kt#L84-L90)).

I'd like to generally avoid `afterEvaluate` and have all interactions and configurations declared up front, especially so users can easily retrieve and do things with them if they want to during configuration time.

**Comment by mkobit** [comment]
Another way I have done it is to use `beforeResolve` and `afterResolve` like I described on this form post (https://discuss.gradle.org/t/configuration-and-transitive-dependency-management-by-extension/23824/5?u=mkobit). That is a little bit more interesting approach but is still a bit wonky.

**Comment by bmuschko** [comment]
Another approach I have been using before in my plugin is to expose a custom configuration (potentially in combination with [assigning a default dependency](https://guides.gradle.org/implementing-gradle-plugins/#providing_default_dependencies_for_plugins)). The drawback is that you have to document the configuration, its purpose and have to assign the full coordinates of dependency.

**Comment by adammurdoch** [comment]
I think this is a good use case: as a plugin author I know I'm going to need a version of something (a tool or library for example) but I don't know exactly what when the plugin is applied. I won't know until some other build logic has configured my model to tell me.

Using a provider here allows me to wire in the dependency when the plugin is applied. I don't need to bother about trying to find some hook that runs after all of the build logic might configure my model but before the dependency is used (aka the hope-for-the-best strategy).

Using a provider is more descriptive, so that we can (later) do things like apply validation to make sure the input value isn't changed after it is used and report on where the dependency came from to make this work better.

We added something similar for declaring the artifacts of the project.

**Comment by bigdaz** [comment]
I'm leaving this `in:core` since it's more about project configuration than `dependency-management`.

**Comment by mkobit** [comment]
This can sort of be worked around better now by using the newer `Configuration.withDependencies` API

**Comment by mkobit** [comment]
Maybe this is less needed with the configuration avoidance APIs? I haven't tried it yet, but:

```kotlin

val myProp: Property<String> = objects.property()
myProp.set("com.google.guava:guava:23.0")

repositories {
  jcenter()
}

val myConfig by configurations.registering

myConfig.configure config@ {
  dependencies {
    this@config(myProp.get())
  }
}
```

**Comment by ov7a** [comment]
Fixed in 6.8
https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/dsl/DependencyHandler.html#addProvider(java.lang.String,org.gradle.api.provider.Provider,org.gradle.api.Action)
